File: permissionsdispatcher-processor/src/test/java/permissions/dispatcher/processor/UtilsTest.java
package permissions.dispatcher.processor;

import org.junit.Test;
import org.mockito.Mockito;

import javax.lang.model.element.ExecutableElement;
import java.lang.annotation.Annotation;
import java.util.Collections;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static permissions.dispatcher.processor.Utils.*;

/**
 * Unit test for {@link Utils}.
 */
public class UtilsTest {

    @Test
    public void testFindShowsRationaleFromValue() {
        List<ExecutableElement> list = Collections.emptyList();
        assertThat(findShowsRationaleFromValue("test", list)).isNull();
    }

    @Test
    public void testFindShowsRationalesFromValue() {
        List<ExecutableElement> list = Collections.emptyList();
        assertThat(findShowsRationalesFromValue(new String[]{"test"}, list)).isNull();
    }

    @Test
    public void testGetValueFromAnnotation() {
        ExecutableElement element = Mockito.mock(ExecutableElement.class);
        assertThat(getValueFromAnnotation(element, Annotation.class)).isEmpty();
    }

    @Test
    public void testGetPackageName() {
        {
            String name = "permissions.dispatcher.Activity";
            assertThat(getPackageName(name)).isEqualTo("permissions.dispatcher");
        }
        {
            String name = "...";
            assertThat(getPackageName(name)).isEqualTo("..");
        }
    }

    @Test
    public void testGetClassName() {
        {
            String name = "permissions.dispatcher.Activity";
            assertThat(getClassName(name)).isEqualTo("Activity");
        }
        {
            String name = "Activity";
            assertThat(getClassName(name)).isEqualTo("Activity");
        }
        {
            String name = "...";
            assertThat(getClassName(name)).isEmpty();
        }
    }

    @Test
    public void testGetFieldName() {
        {
            String name = "activity";
            assertThat(getFieldName(name)).isEqualTo("REQUEST_ACTIVITY");
        }
        {
            String name = "FRAGMENT";
            assertThat(getFieldName(name)).isEqualTo("REQUEST_FRAGMENT");
        }
    }

    @Test
    public void testGetIsEmpty() {
        {
            assertThat(isEmpty(null)).isTrue();
        }
        {
            assertThat(isEmpty(Collections.emptyList())).isTrue();
        }
        {
            assertThat(isEmpty(Collections.singletonList("android"))).isFalse();
        }
    }

    @Test
    public void testToString() {
        {
            assertThat(Utils.toString(null)).isNull();
        }
        {
            assertThat(Utils.toString(new String[]{""})).isEqualTo("{\"\"}");
        }
        {
            assertThat(Utils.toString(new String[]{"android"})).isEqualTo("{\"android\"}");
        }
        {
            assertThat(Utils.toString(new String[]{"android", "ios"})).isEqualTo("{\"android\", \"ios\"}");
        }
    }

}

File: permissionsdispatcher-processor/src/test/java/permissions/dispatcher/processor/UtilsTest.kt
package permissions.dispatcher.processor

import org.junit.Test
import org.mockito.Mockito

import javax.lang.model.element.ExecutableElement
import java.util.Collections

import org.assertj.core.api.Assertions.assertThat

/**
 * Unit test for [Utils].
 */
public class UtilsTest {

    Test
    public fun testFindShowsRationaleFromValue() {
        val list = emptyList<ExecutableElement>()
        assertThat<ExecutableElement>(Utils.findShowsRationaleFromValue("test", list)).isNull()
    }

    Test
    public fun testFindShowsRationalesFromValue() {
        val list = emptyList<ExecutableElement>()
        assertThat(findShowsRationalesFromValue(arrayOf("test"), list)).isNull()
    }

    Test
    public fun testGetValueFromAnnotation() {
        val element = Mockito.mock(javaClass<ExecutableElement>())
        assertThat(getValueFromAnnotation(element, javaClass<Annotation>())).isEmpty()
    }

    Test
    public fun testGetPackageName() {
        run {
            val name = "permissions.dispatcher.Activity"
            assertThat(getPackageName(name)).isEqualTo("permissions.dispatcher")
        }
        run {
            val name = "..."
            assertThat(getPackageName(name)).isEqualTo("..")
        }
    }

    Test
    public fun testGetClassName() {
        run {
            val name = "permissions.dispatcher.Activity"
            assertThat(getClassName(name)).isEqualTo("Activity")
        }
        run {
            val name = "Activity"
            assertThat(getClassName(name)).isEqualTo("Activity")
        }
        run {
            val name = "..."
            assertThat(getClassName(name)).isEmpty()
        }
    }

    Test
    public fun testGetFieldName() {
        run {
            val name = "activity"
            assertThat(getRequestCodeFieldName(name)).isEqualTo("REQUEST_ACTIVITY")
        }
        run {
            val name = "FRAGMENT"
            assertThat(getRequestCodeFieldName(name)).isEqualTo("REQUEST_FRAGMENT")
        }
    }

    Test
    public fun testGetIsEmpty() {
        run { assertThat(isEmpty(null)).isTrue() }
        run { assertThat(isEmpty(emptyList<Any>())).isTrue() }
        run { assertThat(isEmpty(listOf("android"))).isFalse() }
    }

    Test
    public fun testToString() {
        run { //noinspection NullArgumentToVariableArgMethod
            assertThat(Utils.toString(null)).isNull()
        }
        run { assertThat(Utils.toString(*arrayOf(""))).isEqualTo("{\"\"}") }
        run { assertThat(Utils.toString(*arrayOf("android"))).isEqualTo("{\"android\"}") }
        run { assertThat(Utils.toString(*arrayOf("android", "ios"))).isEqualTo("{\"android\", \"ios\"}") }
    }

}
